typescript/src/redemption.ts

```diff
@@ -0,0 +1,299 @@
+  walletPrivateKey: string,
+  mainUtxo: UnspentTransactionOutput,
+  redeemerAddresses: string[],
+  witness: boolean
+): Promise<void> {
+  const rawTransaction = await bitcoinClient.getRawTransaction(
+    mainUtxo.transactionHash
+  )
+
+  const mainUtxoWithRaw: UnspentTransactionOutput & RawTransaction = {
+    ...mainUtxo,
+    transactionHex: rawTransaction.transactionHex,
+  }
+
+  const redemptionRequests = await prepareRedemptionRequests(
+    bridge,
+    walletPrivateKey,
+    redeemerAddresses
+  )
+
+  const transaction = await createRedemptionTransaction(
+    walletPrivateKey,
+    mainUtxoWithRaw,
+    redemptionRequests,
+    witness
+  )
+
+  // Note that `broadcast` may fail silently (i.e. no error will be returned,
+  // even if the transaction is rejected by other nodes and does not enter the
+  // mempool, for example due to an UTXO being already spent).
+  await bitcoinClient.broadcast(transaction)
+}
+
+/**
+ * Creates a Bitcoin redemption transaction.
+ * The transaction will have a single input (main UTXO) and an output for each
+ * redemption request provided and a change output if the redemption requests
+ * do not consume all the Bitcoins from the main UTXO.
+ * @dev The caller is responsible for ensuring the redemption request list is
+ *      correctly formed:
+ *        - there is at least one redemption
+ *        - the `requestedAmount` in each redemption request is greater than
+ *          the sum of its `feeShare` and `treasuryFee`
+ *        - the redeemer address in each redemption request is of a standard
+ *          type (P2PKH, P2WPKH, P2SH, P2WSH).
+ * @param walletPrivateKey  - The private key of the wallet in the WIF format
+ * @param mainUtxo - The main UTXO of the wallet. Must be P2(W)PKH
+ * @param redemptionRequests - The list of redemption requests
+ * @param witness - The parameter used to decide the type of the change output.
+ *                  P2WPKH if `true`, P2PKH if `false`
+ * @returns Bitcoin redemption transaction in the raw format.
+ */
+export async function createRedemptionTransaction(
+  walletPrivateKey: string,
+  mainUtxo: UnspentTransactionOutput & RawTransaction,
+  redemptionRequests: RedemptionRequest[],
+  witness: boolean
+): Promise<RawTransaction> {
+  if (redemptionRequests.length < 1) {
+    throw new Error("There must be at least one request to redeem")
+  }
+
+  const decodedWalletPrivateKey = wif.decode(walletPrivateKey)
+
+  const walletKeyRing = new bcoin.KeyRing({
+    witness: witness,
+    privateKey: decodedWalletPrivateKey.privateKey,
+    compressed: decodedWalletPrivateKey.compressed,
+  })
+
+  const walletAddress = walletKeyRing.getAddress("string")
+
+  // Use the main UTXO as the single transaction input
+  const inputCoins = [
+    bcoin.Coin.fromTX(
+      bcoin.MTX.fromRaw(mainUtxo.transactionHex, "hex"),
+      mainUtxo.outputIndex,
+      -1
+    ),
+  ]
+
+  const transaction = new bcoin.MTX()
+
+  let txFee = 0
+  let totalOutputValue = 0
```

Comment:
`totalOutputsValue`?

Reply:
Done in 0e7002cd00ddd49533ead638b3dd42c0fe81ef47.